File: solutions/go/roman-numerals/1/roman_numerals.go
package romannumerals

import "errors"

type Pair struct {
	decimal int
	roman   string
}

var valueSymbols = []Pair{
	{1000, "M"},
	{900, "CM"},
	{500, "D"},
	{400, "CD"},
	{100, "C"},
	{90, "XC"},
	{50, "L"},
	{40, "XL"},
	{10, "X"},
	{9, "IX"},
	{5, "V"},
	{4, "IV"},
	{1, "I"},
}

func ToRomanNumeral(input int) (string, error) {
	romanNumbers := ""
	if input < 1 || input > 3999 {
		return romanNumbers, errors.New("invalid number")
	}
	remainder := input
	for _, pair := range valueSymbols {
		for {
			if remainder < pair.decimal {
				break
			}
			romanNumbers += pair.roman
			remainder -= pair.decimal
		}
	}
	return romanNumbers, nil
}

File: solutions/java/roman-numerals/1/src/main/java/RomanNumerals.java
class RomanNumerals {
    private int number;
    
    RomanNumerals(int number) {
        this.number = number;
    }

    String getRomanNumeral() {
        StringBuilder resultBuilder = new StringBuilder();
        while (this.number > 0) {
            if (this.number >= 1000) {
                resultBuilder.append("M"); 
                this.number -= 1000;                  
            }
            else if (this.number >= 900) {
                resultBuilder.append("CM"); 
                this.number -= 900;                  
            }
            else if (this.number >= 500) {
                resultBuilder.append("D"); 
                this.number -= 500;                  
            }
            else if (this.number >= 400) {
                resultBuilder.append("CD"); 
                this.number -= 400;                  
            }
            else if (this.number >= 100) {
                resultBuilder.append("C"); 
                this.number -= 100;                  
            }
            else if (this.number >= 90) {
                resultBuilder.append("XC"); 
                this.number -= 90;                  
            }
            else if (this.number >= 50) {
                resultBuilder.append("L"); 
                this.number -= 50;                  
            }
            else if (this.number >= 40) {
                resultBuilder.append("XL"); 
                this.number -= 40;                  
            }
            else if (this.number >= 10) {
                resultBuilder.append("X"); 
                this.number -= 10;                  
            }
            else if (this.number >= 9) {
                resultBuilder.append("IX"); 
                this.number -= 9;                  
            }
            else if (this.number >= 5) {
                resultBuilder.append("V"); 
                this.number -= 5;                  
            }
            else if (this.number >= 4) {
                resultBuilder.append("IV"); 
                this.number -= 4;                  
            }
            else if (this.number >= 1) {
                resultBuilder.append("I"); 
                this.number -= 1;                  
            }
        }
        return resultBuilder.toString();
    }
}
